dags/max_ai/utils.py

```diff
@@ -0,0 +1,85 @@
+import hashlib
+from collections.abc import Sequence
+from contextlib import contextmanager
+from datetime import datetime
+from tempfile import TemporaryFile
+
+import botocore
+from dagster_aws.s3 import S3Resource
+from django.conf import settings
+from fastavro import parse_schema, writer
+from pydantic_avro import AvroBase
+
+EVALS_S3_PREFIX = "ai_evals"
+
+# objectstorage has only the default bucket in debug.
+if settings.DEBUG:
+    EVALS_S3_BUCKET = settings.OBJECT_STORAGE_BUCKET
+else:
+    EVALS_S3_BUCKET = settings.DAGSTER_AI_EVALS_S3_BUCKET
+
+
+def get_consistent_hash_suffix(file_name: str, date: datetime | None = None, code_version: str | None = None) -> str:
+    """
+    Generate a consistent hash suffix that updates twice per month based on the filename.
+
+    The hash changes on the 1st and 15th of each month, ensuring links update
+    twice monthly while remaining consistent within each period.
+
+    Args:
+        file_name: The base filename to hash
+        date: Optional date for testing, defaults to current date
+        code_version: Optional code version for hash consistency
+
+    Returns:
+        A short hash string (8 characters) that's consistent within each half-month period
+    """
+    if date is None:
+        date = datetime.now()
+
+    # Determine which half of the month we're in
+    half_month_period = 1 if date.day < 15 else 2
+
+    # Create a seed that changes twice per month
+    period_seed = f"{date.year}-{date.month:02d}-{half_month_period}"
+
+    # Combine the period seed with the filename for consistent hashing
+    hash_input = f"{period_seed}:{file_name}"
+    if code_version:
+        hash_input += f":{code_version}"
+
+    # Generate a short, URL-safe hash
+    hash_obj = hashlib.sha256(hash_input.encode("utf-8"))
+    return hash_obj.hexdigest()[:8]
+
+
+def compose_postgres_dump_path(project_id: int, file_name: str, code_version: str | None = None) -> str:
+    """Compose S3 path for Postgres dumps with consistent hashing"""
+    hash_suffix = get_consistent_hash_suffix(file_name, code_version=code_version)
+    versioned_file_name = f"{file_name}_{hash_suffix}"
+    return f"{EVALS_S3_PREFIX}/postgres_models/{project_id}/{versioned_file_name}.avro"
```

Comment:
What do you think about the `filename` being actually a dir? Instead of it being included in the `versioned_file_name`. Maybe it helps our dev experience when we debug or when we wanna check what type of models are we pushing per each project. If I wanna search for a file I can simply check the dir of the model lets say `property_definitions` and then find the most recent there. 
```suggestion
    return f"{EVALS_S3_PREFIX}/postgres_models/{project_id}/{file_name}/{hash_suffix}.avro"
```

Reply:
It is actually already included and looks like this:

```
event_taxonomy
ai_evals/clickhouse_queries/1/events_taxonomy_e5c725bc.avro
properties_taxonomy
ai_evals/clickhouse_queries/1/properties_taxonomy_ac9ab2a3.avro
actors_property_taxonomy
ai_evals/clickhouse_queries/1/actors_property_taxonomy_1f4c9db0.avro
```